### PR TITLE
Add support for environment map transformation

### PR DESCRIPTION
**Pull request by Soulghost**
# Objective

- Fixes: https://github.com/bevyengine/bevy/issues/14036

## Solution

- Add a world space transformation for the environment sample direction.

## Testing

- I have tested the newly added `transform` field using the newly added `rotate_environment_map` example. 

https://github.com/user-attachments/assets/2de77c65-14bc-48ee-b76a-fb4e9782dbdb

## Migration Guide

- Since we have added a new filed to the `EnvironmentMapLight` struct, users will need to include `..default()` or some rotation value in their initialization code.

### CODE REVIEW

File: crates/bevy_pbr/src/light_probe/environment_map.rs [hunk 1]
```diff
@@ -96,6 +101,34 @@ pub struct EnvironmentMapLight {
     ///
     /// See also <https://google.github.io/filament/Filament.html#lighting/imagebasedlights/iblunit>.
     pub intensity: f32,
+
+    /// The rotation of the environment light map
```
Comment (alice-i-cecile) [inline crates/bevy_pbr/src/light_probe/environment_map.rs:105]: Nit: Documenting how this rotation is applied would be useful to reduce trial-and-error here.

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl [hunk 2]
```diff
@@ -57,17 +58,21 @@ fn compute_radiances(
         bindings::specular_environment_maps[query_result.texture_index]) - 1u);
 
     if (!found_diffuse_indirect) {
+        var irradiance_sample_dir = vec3(N.xy, -N.z);
+        irradiance_sample_dir = (environment_map_uniform.transform * vec4(irradiance_sample_dir, 1.0)).xyz;
```
Comment (alice-i-cecile) [inline crates/bevy_pbr/src/light_probe/environment_map.wgsl:62]: A comment here (and below) would be useful to make this math easier to follow.

File: crates/bevy_pbr/src/light_probe/mod.rs [hunk 3]
```diff
@@ -296,6 +297,11 @@ impl LightProbe {
     }
 }
 
+#[derive(Component, ShaderType, Clone)]
+pub struct EnvironmentMapUniform {
```
Comment (alice-i-cecile) [inline crates/bevy_pbr/src/light_probe/mod.rs:301]: Documentation please: this is a public type.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Useful feature, but I'd like to see more documentation and a use of this in an example to make it easier to test.

**Comment by Azorlogh** [comment]
I think the rotation is backward compared to the other skybox rotation PR:

This is what I get when the skybox & env maps share the same rotation (notice we see what's in front of us in the reflection instead of what's behind)

https://github.com/user-attachments/assets/634a2194-f327-4612-b75f-00396be6c4b4

**Comment by Soulghost** [comment]
> I think the rotation is backward compared to the other skybox rotation PR:
> 
> This is what I get when the skybox & env maps share the same rotation (notice we see what's in front of us in the reflection instead of what's behind)
> 
>  1721060695.mp4

Yes, I have reproduced this bug using a debug texture, which more clearly illustrates this issue:
panorama_image

The result indicates that the skybox and the environment map are rotated in opposite directions:
image

I found some clues here: "Cube maps are left-handed so we negate the z coordinate" (https://www.khronos.org/opengl/wiki/Cubemap_Texture)
```wgsl
@fragment
fn skybox_fragment(in: VertexOutput) -> @location(0) vec4<f32> {
    let ray_direction = coords_to_ray_direction(in.position.xy, view.viewport);

    // Cube maps are left-handed so we negate the z coordinate.
    let out = textureSample(skybox, skybox_sampler, ray_direction * vec3(1.0, 1.0, -1.0));
    return vec4(out.rgb * uniforms.brightness, out.a);
}
```

For the skybox cubemap, we rotate the ray in a right-handed view space coordinate, while for the environment map, we rotate it in a right-handed world space coordinate. However, there appears to be a difference between these two types of cubemaps. I will do some research to figure it out.

**Comment by Soulghost** [comment]
I've identified the issue: In the environment map, I initially inverted the z before transforming, which placed the transformation in a left-handed coordinate system. I should transform first, like with the skybox, and then invert the z. 😄

**Comment by Hexorg** [comment]
Great catch!

**Review by alice-i-cecile — APPROVED** [review]

**Review by Azorlogh — APPROVED** [review]

**Comment by Soulghost** [comment]
There appears to be some binding errors for the `alien_cake_addict` example, let me check where the problem is.

File: crates/bevy_pbr/src/render/mesh_view_bindings.rs [hunk 4]
```diff
@@ -484,6 +489,7 @@ pub fn prepare_mesh_view_bind_groups(
         Some(light_probes_binding),
         Some(visibility_ranges_buffer),
         Some(ssr_binding),
+        Some(environment_map_binding),
```
Comment (Soulghost) [inline crates/bevy_pbr/src/render/mesh_view_bindings.rs:492]: Will exit if no `EnvironmentMapLight ` is present in the scene, need to implement a fallback uniform for the `environment_map_binding`

**Comment by Soulghost** [comment]
@alice-i-cecile @Azorlogh I apologize for the mistake that led to the mismatch of the mesh view bindings. I should have conducted more thorough testing before submiting these changes. I have now resolved all the issues and tested numerous examples. Below is a summary of the code changes.

1. The `EnvironmentMapUniform` should be configured as a per-view buffer, therefore I have modified the uniform to use a dynamic offset. We also need to include the `ViewEnvironmentMapUniformOffset` in all the associated `ViewNode`s and `RenderCommand`s.
    ```rust
    uniform_buffer::<EnvironmentMapUniform>(true).visibility(ShaderStages::FRAGMENT)
    ```

2. Some of the camera may not contain a `EnvironmentMapLight`, so we need to fallback to a default binding for the `EnvironmentMapUniform`
    ```rust
    fn gather_environment_map_uniform(
        view_query: Extract<Query<(Entity, Option<&EnvironmentMapLight>), With<Camera3d>>>,
        mut commands: Commands,
    ) {
        for (view_entity, environment_map_light) in view_query.iter() {
            let environment_map_uniform = if let Some(environment_map_light) = environment_map_light {
                EnvironmentMapUniform {
                    transform: Transform::from_rotation(environment_map_light.rotation)
                        .compute_matrix()
                        .inverse(),
                }
            } else {
                EnvironmentMapUniform::default()
            };
            commands
                .get_or_spawn(view_entity)
                .insert(environment_map_uniform);
        }
    }
    ```

3. Changing the binding number in `mesh_view_bindings.wgsl` and `mesh_view_bindings.rs` is insufficient; the tone mapping LUT bindings also required adjustment. Since these numbers are injected through shader definitions, they are easily overloooked. **This is the root cause of this issue**.
    ```rust
    #define_import_path bevy_core_pipeline::tonemapping_lut_bindings
    
    @group(0) @binding(#TONEMAPPING_LUT_TEXTURE_BINDING_INDEX) var dt_lut_texture: texture_3d<f32>;
    @group(0) @binding(#TONEMAPPING_LUT_SAMPLER_BINDING_INDEX) var dt_lut_sampler: sampler;
    ```
    
    ```
    diff --git a/crates/bevy_pbr/src/render/mesh.rs b/crates/bevy_pbr/src/render/mesh.rs
    index 8432cbd97..282b116e9 100644
    --- a/crates/bevy_pbr/src/render/mesh.rs
    +++ b/crates/bevy_pbr/src/render/mesh.rs
    @@ -1765,11 +1765,11 @@ impl SpecializedMeshPipeline for MeshPipeline {
                 shader_defs.push("TONEMAP_IN_SHADER".into());
                 shader_defs.push(ShaderDefVal::UInt(
                     "TONEMAPPING_LUT_TEXTURE_BINDING_INDEX".into(),
    -                20,
    +                21,
                 ));
                 shader_defs.push(ShaderDefVal::UInt(
                     "TONEMAPPING_LUT_SAMPLER_BINDING_INDEX".into(),
    -                21,
    +                22,
                 ));
    ```

**Review by alice-i-cecile — APPROVED** [review]

**Pull request merged**